File: 박상혁/WEEK4_계단오르기.cpp
#include <iostream>
#define N_MAX 300

using namespace std;

int stairs[N_MAX + 1];
// memo[i][0]은 직전 칸을 밟지 않음. memo[i][1]은 직전 칸을 밟음
int memo[N_MAX + 1][2];

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0); cout.tie(0);

	int N;
	cin >> N;
	for (int i = 1; i <= N; ++i)
	{
		cin >> stairs[i];
	}

	memo[1][0] = memo[1][1] = stairs[1];
	
	for (int i = 2; i <= N; ++i)
	{
		memo[i][0] = max(memo[i - 2][0], memo[i - 2][1]) + stairs[i];
		memo[i][1] = memo[i - 1][0] + stairs[i];
	}

	cout << max(memo[N][0], memo[N][1]);

	return 0;
}

File: 김성연/WEEK4 계단오르기.py
n = int(input())
stairs = [int(input()) for _ in range(n)]

if n <= 2:
    print(sum(stairs))
else:
    dp = [0] * n
    dp[0], dp[1] = stairs[0], stairs[0] + stairs[1] # 초기값 설정

    for i in range(2, n):
        dp[i] = max(dp[i-3]+stairs[i-1]+stairs[i], dp[i-2]+stairs[i])

    print(dp[-1])
